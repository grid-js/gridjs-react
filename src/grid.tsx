import * as React from "react";
import { Component, createRef, RefObject } from "react";
import { Grid as Gridjs, UserConfig } from "gridjs";

class Grid extends Component<Partial<UserConfig>, any> {
  private wrapper: RefObject<HTMLDivElement> = createRef();
  // Grid.js instance
  private readonly instance = null;

  constructor(props) {
    super(props);

    this.instance = new Gridjs(props || {});
  }

  getInstance(): Gridjs {
    return this.instance;
  }

  componentDidMount() {
    if (this.wrapper.current) this.instance.render(this.wrapper.current);
  }

  componentDidUpdate() {
    if (this.wrapper.current) this.instance.updateConfig(this.props).forceRender();
  }

  render() {
    return <div ref={this.wrapper} />;
  }
}

export default Grid;
