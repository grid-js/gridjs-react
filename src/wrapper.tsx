import { h, createRef as gCreateRef, Component as gComponent } from "gridjs";
import * as ReactDOMClient from 'react-dom/client';

export class ReactWrapper extends gComponent<{
  element: any;
  parent?: string;
}> {
  static defaultProps = {
    parent: "div",
  };

  ref = gCreateRef();

  componentDidMount() {
    const root = ReactDOMClient.createRoot(this.ref.current);
    root.render(this.props.element);
  }

  render() {
    return h(this.props.parent, { ref: this.ref });
  }
}

export function _(element: any, parent?: string) {
  return h(ReactWrapper, {
    element: element,
    parent: parent,
  });
}
