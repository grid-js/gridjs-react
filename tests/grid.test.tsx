import * as React from "react";
import {render, waitFor} from '@testing-library/react'
import { Grid } from "../index";
import { _, ReactWrapper } from "../src/wrapper";
import { h } from "gridjs";


describe("Grid component", () => {
  it("should render a table without header", async () => {
    const table = render(<Grid data={[[1, 2, 3]]} />);

    await waitFor(() => table.findByText('1'));

    expect(table.container.innerHTML).toMatchSnapshot();
  });

  it("should render a table with headers", async () => {
    const table = render(
      <Grid data={[[1, 2, 3]]} columns={["a", "b", "c"]} />
    );

    await waitFor(() => table.findByText('1'));

    expect(table.container.innerHTML).toMatchSnapshot();
  });

  //it("should receive the ready event", async () => {
  //  const table = render(
  //    <Grid data={[[1, 2, 3]]} columns={["a", "b", "c"]} width={"500px"} />
  //  );

  //  const fn = jest.fn();
  //  instance(table).on('ready', fn);

  //  table.update();
  //  await flushPromises();

  //  expect(fn).toBeCalledTimes(1);
  //});

  it("should render a table with width", async () => {
    const table = render(
      <Grid data={[[1, 2, 3]]} columns={["a", "b", "c"]} width={"500px"} />
    );

    await waitFor(() => table.findByText('1'));

    expect(table.container.innerHTML).toMatchSnapshot();
  });

  it("should render a table with search", async () => {
    const table = render(
      <Grid data={[[1, 2, 3]]} columns={["a", "b", "c"]} search={true} />
    );

    await waitFor(() => table.findByText('1'));
    expect(table.container.innerHTML).toMatchSnapshot();
  });

  it("should render a table with pagination and search", async () => {
    const table = render(
      <Grid
        data={[
          [10, 20, 30],
          [40, 50, 60],
          [70, 80, 90],
        ]}
        columns={["a", "b", "c"]}
        search={true}
        pagination={{
          limit: 1,
        }}
      />
    );

    await waitFor(() => table.findByText('30'));

    expect(table.container.innerHTML).toMatchSnapshot();
  });

  it("should render a table with the ReactWrapper class", async () => {
    const table = render(
      <Grid
        data={[
          [1, 2, 3],
          [4, 5, 6],
          [
            h(ReactWrapper, {
              element: <b>7</b>,
            }),
            h(ReactWrapper, {
              element: <b>8</b>,
            }),
            h(ReactWrapper, {
              element: <b>9</b>,
              parent: "span",
            }),
          ],
        ]}
        columns={["a", "b", "c"]}
      />
    );

    await waitFor(() => table.findByText('8'));

    expect(table.container.innerHTML).toMatchSnapshot();
  });

  it("should render a table with the wrapper function", async () => {
    const table = render(
      <Grid
        data={[
          [1, 2, 3],
          [4, 5, 6],
          [_(<b>7</b>), _(<b>8</b>), _(<span>9</span>, "span")],
        ]}
        columns={["a", "b", "c"]}
      />
    );

    await waitFor(() => table.findByText('9'));

    expect(table.container.innerHTML).toMatchSnapshot();
  });
});
