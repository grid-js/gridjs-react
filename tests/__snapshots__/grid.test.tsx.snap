// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid component should render a table with headers 1`] = `"<div><div role="complementary" class="gridjs gridjs-container" style="width: 100%;"><div class="gridjs-wrapper" style="height: auto;"><table role="grid" class="gridjs-table" style="height: auto;"><thead class="gridjs-thead"><tr class="gridjs-tr"><th data-column-id="a" class="gridjs-th"><div class="gridjs-th-content">a</div></th><th data-column-id="b" class="gridjs-th"><div class="gridjs-th-content">b</div></th><th data-column-id="c" class="gridjs-th"><div class="gridjs-th-content">c</div></th></tr></thead><tbody class="gridjs-tbody"><tr class="gridjs-tr"><td data-column-id="a" class="gridjs-td">1</td><td data-column-id="b" class="gridjs-td">2</td><td data-column-id="c" class="gridjs-td">3</td></tr></tbody></table></div><div id="gridjs-temp" class="gridjs-temp"></div></div></div>"`;

exports[`Grid component should render a table with pagination and search 1`] = `"<div><div role="complementary" class="gridjs gridjs-container" style="width: 100%;"><div class="gridjs-head"><div class="gridjs-search"><input type="search" placeholder="Type a keyword..." aria-label="Type a keyword..." class="gridjs-input gridjs-search-input"></div></div><div class="gridjs-wrapper" style="height: auto;"><table role="grid" class="gridjs-table" style="height: auto;"><thead class="gridjs-thead"><tr class="gridjs-tr"><th data-column-id="a" class="gridjs-th"><div class="gridjs-th-content">a</div></th><th data-column-id="b" class="gridjs-th"><div class="gridjs-th-content">b</div></th><th data-column-id="c" class="gridjs-th"><div class="gridjs-th-content">c</div></th></tr></thead><tbody class="gridjs-tbody"><tr class="gridjs-tr"><td data-column-id="a" class="gridjs-td">10</td><td data-column-id="b" class="gridjs-td">20</td><td data-column-id="c" class="gridjs-td">30</td></tr></tbody></table></div><div class="gridjs-footer"><div class="gridjs-pagination"><div role="status" aria-live="polite" class="gridjs-summary" title="Page 1 of 3">Showing <b>1</b> to <b>1</b> of <b>3</b> results</div><div class="gridjs-pages"><button tabindex="0" role="button" disabled="" title="Previous" aria-label="Previous" class="">Previous</button><button tabindex="0" role="button" class="gridjs-currentPage" title="Page 1" aria-label="Page 1">1</button><button tabindex="0" role="button" class="" title="Page 2" aria-label="Page 2">2</button><button tabindex="0" role="button" class="" title="Page 3" aria-label="Page 3">3</button><button tabindex="0" role="button" title="Next" aria-label="Next" class="">Next</button></div></div></div><div id="gridjs-temp" class="gridjs-temp"></div></div></div>"`;

exports[`Grid component should render a table with search 1`] = `"<div><div role="complementary" class="gridjs gridjs-container" style="width: 100%;"><div class="gridjs-head"><div class="gridjs-search"><input type="search" placeholder="Type a keyword..." aria-label="Type a keyword..." class="gridjs-input gridjs-search-input"></div></div><div class="gridjs-wrapper" style="height: auto;"><table role="grid" class="gridjs-table" style="height: auto;"><thead class="gridjs-thead"><tr class="gridjs-tr"><th data-column-id="a" class="gridjs-th"><div class="gridjs-th-content">a</div></th><th data-column-id="b" class="gridjs-th"><div class="gridjs-th-content">b</div></th><th data-column-id="c" class="gridjs-th"><div class="gridjs-th-content">c</div></th></tr></thead><tbody class="gridjs-tbody"><tr class="gridjs-tr"><td data-column-id="a" class="gridjs-td">1</td><td data-column-id="b" class="gridjs-td">2</td><td data-column-id="c" class="gridjs-td">3</td></tr></tbody></table></div><div id="gridjs-temp" class="gridjs-temp"></div></div></div>"`;

exports[`Grid component should render a table with the ReactWrapper class 1`] = `"<div><div role="complementary" class="gridjs gridjs-container" style="width: 100%;"><div class="gridjs-wrapper" style="height: auto;"><table role="grid" class="gridjs-table" style="height: auto;"><thead class="gridjs-thead"><tr class="gridjs-tr"><th data-column-id="a" class="gridjs-th"><div class="gridjs-th-content">a</div></th><th data-column-id="b" class="gridjs-th"><div class="gridjs-th-content">b</div></th><th data-column-id="c" class="gridjs-th"><div class="gridjs-th-content">c</div></th></tr></thead><tbody class="gridjs-tbody"><tr class="gridjs-tr"><td data-column-id="a" class="gridjs-td">1</td><td data-column-id="b" class="gridjs-td">2</td><td data-column-id="c" class="gridjs-td">3</td></tr><tr class="gridjs-tr"><td data-column-id="a" class="gridjs-td">4</td><td data-column-id="b" class="gridjs-td">5</td><td data-column-id="c" class="gridjs-td">6</td></tr><tr class="gridjs-tr"><td data-column-id="a" class="gridjs-td"><div><b>7</b></div></td><td data-column-id="b" class="gridjs-td"><div><b>8</b></div></td><td data-column-id="c" class="gridjs-td"><span><b>9</b></span></td></tr></tbody></table></div><div id="gridjs-temp" class="gridjs-temp"></div></div></div>"`;

exports[`Grid component should render a table with the wrapper function 1`] = `"<div><div role="complementary" class="gridjs gridjs-container" style="width: 100%;"><div class="gridjs-wrapper" style="height: auto;"><table role="grid" class="gridjs-table" style="height: auto;"><thead class="gridjs-thead"><tr class="gridjs-tr"><th data-column-id="a" class="gridjs-th"><div class="gridjs-th-content">a</div></th><th data-column-id="b" class="gridjs-th"><div class="gridjs-th-content">b</div></th><th data-column-id="c" class="gridjs-th"><div class="gridjs-th-content">c</div></th></tr></thead><tbody class="gridjs-tbody"><tr class="gridjs-tr"><td data-column-id="a" class="gridjs-td">1</td><td data-column-id="b" class="gridjs-td">2</td><td data-column-id="c" class="gridjs-td">3</td></tr><tr class="gridjs-tr"><td data-column-id="a" class="gridjs-td">4</td><td data-column-id="b" class="gridjs-td">5</td><td data-column-id="c" class="gridjs-td">6</td></tr><tr class="gridjs-tr"><td data-column-id="a" class="gridjs-td"><div><b>7</b></div></td><td data-column-id="b" class="gridjs-td"><div><b>8</b></div></td><td data-column-id="c" class="gridjs-td"><span><span>9</span></span></td></tr></tbody></table></div><div id="gridjs-temp" class="gridjs-temp"></div></div></div>"`;

exports[`Grid component should render a table with width 1`] = `"<div><div role="complementary" class="gridjs gridjs-container" style="width: 500px;"><div class="gridjs-wrapper" style="height: auto;"><table role="grid" class="gridjs-table" style="height: auto;"><thead class="gridjs-thead"><tr class="gridjs-tr"><th data-column-id="a" class="gridjs-th"><div class="gridjs-th-content">a</div></th><th data-column-id="b" class="gridjs-th"><div class="gridjs-th-content">b</div></th><th data-column-id="c" class="gridjs-th"><div class="gridjs-th-content">c</div></th></tr></thead><tbody class="gridjs-tbody"><tr class="gridjs-tr"><td data-column-id="a" class="gridjs-td">1</td><td data-column-id="b" class="gridjs-td">2</td><td data-column-id="c" class="gridjs-td">3</td></tr></tbody></table></div><div id="gridjs-temp" class="gridjs-temp"></div></div></div>"`;

exports[`Grid component should render a table without header 1`] = `"<div><div role="complementary" class="gridjs gridjs-container" style="width: 100%;"><div class="gridjs-wrapper" style="height: auto;"><table role="grid" class="gridjs-table" style="height: auto;"><tbody class="gridjs-tbody"><tr class="gridjs-tr"><td class="gridjs-td">1</td><td class="gridjs-td">2</td><td class="gridjs-td">3</td></tr></tbody></table></div><div id="gridjs-temp" class="gridjs-temp"></div></div></div>"`;
